jupyter_server_fileid/handlers.py

```diff
@@ -0,0 +1,21 @@
+from jupyter_server.auth import authorized
+from jupyter_server.base.handlers import APIHandler
+from tornado import web
+
+AUTH_RESOURCE = "contents"
+
+
+class FileIdAPIHandler(APIHandler):
+    auth_resource = AUTH_RESOURCE
+
+
+class FileIdHandler(FileIdAPIHandler):
+    @web.authenticated
+    @authorized
+    async def get(self, path):
+        idx = self.settings["file_id_manager"].index(path)
```

Comment:
Shouldn't this be: 
```suggestion
        idx = self.settings["file_id_manager"].get_id(path)
```
otherwise, `index()` will add the path (assuming it can be `stat`'d) and return the newly created ID.

Are you planning on supporting the ability to fetch a file's path given its ID?  This will be necessary for external references (e.g., comments) to locate their parent/host file.

Reply:
It seems that `get_id` doesn't index the file automatically, while `index` does. So I guess we need the latter?

Reply:
> Are you planning on supporting the ability to fetch a file's path given its ID?

Maybe in another PR, but it's not needed for now in JupyterLab.

Reply:
> It seems that get_id doesn't index the file automatically, while index does. So I guess we need the latter?

I think so - otherwise you're potentially performing a transaction on a fetch request - which is contrary to REST best practices (AFAIU).

> Maybe in another PR, but it's not needed for now in JupyterLab.

Ok - this should be fine.  You won't be able to use a similar format of endpoint in that case as it will be ambiguous so I wasn't sure if supporting path-fetching will affect _this_ endpoint.  You could probably take an approach similar to [kernel interrupt and restarts](https://github.com/jupyter-server/jupyter_server/blob/1957aa0416c8467961198fa2dd6f930aca106f85/jupyter_server/services/kernels/handlers.py#L801-L808) with an endpoint definition of something like `api/fileid/{id,path}/(.*)` although I'm sure there are more elegant approaches.  My comment was more about not impacting what is essentially "set in stone" when fetching paths is required - which will be the case.

Reply:
> I think so - otherwise you're potentially performing a transaction on a fetch request - which is contrary to REST best practices (AFAIU).

I think you misunderstood me, we need `index` which might perform a transaction.

> with an endpoint definition of something like `api/fileid/{id,path}/(.*)`

I think you're right, we need `api/fileid/id/(.*)` here, and we'll need `api/fileid/path/(.*)` when we want to get the path corresponding to a given ID.

Reply:
Agree that we will need both the path and id variant. @dlqqq can you help us understand which underlying methods to call?

Reply:
I think Kevin's concern is that a GET request shouldn't be creating a new resource under RESTful semantics. Indexing creates a file ID "resource" and is also an idempotent operation, so under RESTful semantics, it should be called via PUT.

An API that conforms strictly to RESTful semantics might look something like this:

```
GET /api/fileid/files/{path or id} => calls get_path() if given ID, or get_id() if given path
PUT /api/fileid/ids => index a file at a given path (provided in request payload or URI query string)
```

An API compatible with @davidbrochart's suggestion might look something like this:

```
GET /api/fileid/{path or id} => index a file if given path, or get_path() if given ID
```

Personally, I'm inclined towards Kevin's suggestion as it allows REST API users to distinguish between calling `get_id()` and `index()`, which is useful in scenarios where a user wants to check whether a file was indexed previously but doesn't want to index it now.

Reply:
The obvious issue with putting the path or ID in the request URL rather than as a query string is that it's not possible for the backend to distinguish whether you're talking about a file with the name of a UUID located in the server root or if you're talking about a file ID. So regardless of which high-level API design we settle on, we should probably modify the GET requests to explicitly specify whether they're passing in a path or an ID via `?path=...` or `?id=...`.

Reply:
On the other hand, from the client point of view, GETting a file ID seems like the right HTTP verb. The server creating a new resource is an implementation detail. For instance, if all files under a root directory were indexed at startup and then tracked for any change (renaming, etc.), then there would be no need for indexing in this GET request. The client doesn't ask for a new resource, it asks for a property of a file.
I don't have any preference as to where to specify what we request (an ID for a path or a path for an ID). It could be in the URL path:
```
GET /api/fileid/id/{path}: get an ID for a path
GET /api/fileid/path/{id}: get a path for an ID
```
or as a query parameter:
```
GET /api/fileid?path={path}: get an ID for a path
GET /api/fileid?id={id}: get a path for an ID
```
The latter is slightly more complicated as we need to check that only one parameter is passed, and the returned schema could theoretically be different for a path or an ID, but it's not a big deal.